File: src/32_01_PrintTreeFromTopToBottom/32_01_PrintTreeFromTopToBottom.cpp
// 面试题32（一）：不分行从上往下打印二叉树
// 题目：从上往下打印出二叉树的每个结点，同一层的结点按照从左到右的顺序打印。

#include <iostream>
#include <queue>
#include <string>
#include <vector>

#include "../Utilities/BinaryTree.h"

using namespace std;

using TreeNode = BinaryTreeNode;

class Solution {
   public:
    // 使用队列实现BFS
    vector<int> levelOrder( TreeNode* root ) {
        vector<int> results;
        if ( !root ) return results;

        queue<TreeNode*> q;
        q.push( root );
        while ( !q.empty() ) {
            TreeNode* tmp = q.front();
            q.pop();
            results.push_back( tmp->val );
            if ( tmp->left ) q.push( tmp->left );
            if ( tmp->right ) q.push( tmp->right );
        }
        return results;
    }
};

class Test {
   public:
    void checkResult( const string testName, TreeNode* root ) {
        cout << testName << ": [ ";
        vector<int> result = solution.levelOrder( root );
        for ( int r : result ) cout << r << ", ";
        cout << "]\n";
    }

    //            10
    //         /      \
    //        6        14
    //       /\        /\
    //      4  8     12  16
    void Test1() {
        ValVec vals{ 10, 6, 14, 4, 8, 12, 16 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, 1, 2 );
        connectNodesIndex( nodes, 1, 3, 4 );
        connectNodesIndex( nodes, 2, 5, 6 );

        checkResult( "Test1", nodes[ 0 ] );
        destroyTree( nodes[ 0 ] );
    }

    //               5
    //              /
    //             4
    //            /
    //           3
    //          /
    //         2
    //        /
    //       1
    void Test2() {
        ValVec vals{ 5, 4, 3, 2, 1 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, 1, -1 );
        connectNodesIndex( nodes, 1, 2, -1 );
        connectNodesIndex( nodes, 2, 3, -1 );
        connectNodesIndex( nodes, 3, 4, -1 );

        checkResult( "Test2", nodes[ 0 ] );
        destroyTree( nodes[ 0 ] );
    }

    // 1
    //  \
    //   2
    //    \
    //     3
    //      \
    //       4
    //        \
    //         5
    void Test3() {
        ValVec vals{ 1, 2, 3, 4, 5 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, -1, 1 );
        connectNodesIndex( nodes, 1, -1, 2 );
        connectNodesIndex( nodes, 2, -1, 3 );
        connectNodesIndex( nodes, 3, -1, 4 );

        checkResult( "Test3", nodes[ 0 ] );
        destroyTree( nodes[ 0 ] );
    }

    // 树中只有1个结点
    void Test4() {
        ValVec vals{ 1 };
        NodeVec nodes = createNodesVec( vals );

        checkResult( "Test4", nodes[ 0 ] );
        destroyTree( nodes[ 0 ] );
    }

    // 树中没有结点
    void Test5() { checkResult( "Test5", nullptr ); }

    void runTest() {
        Test1();
        Test2();
        Test3();
        Test4();
        Test5();
    }

   private:
    Solution solution;
};

int main() {
    Test test;
    test.runTest();
    return 0;
}

File: src/32_01_PrintTreeFromTopToBottom/32_01_PrintTreeFromTopToBottom.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import os
import sys

sys.path.append(os.path.split(sys.path[0])[0])


from Utilities import binary_tree
from Utilities.binary_tree import TreeNode


class Solution:
    def levelOrder(self, root: TreeNode) -> list[int]:
        if not root:
            return []
        results, q = [], [root]
        while q:
            tmp = q.pop(0)
            results.append(tmp.val)
            if tmp.left:
                q.append(tmp.left)
            if tmp.right:
                q.append(tmp.right)
        return results


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(self, testName: str, root: TreeNode):
        text = testName
        result = self.solution.levelOrder(root)
        text += f"{result}"
        print(text)

    #            10
    #         /      \
    #        6        14
    #       /\        /\
    #      4  8     12  16
    def Test1(self):
        vals = [10, 6, 14, 4, 8, 12, 16]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, 1, 2),
            (1, 3, 4),
            (2, 5, 6),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test1", nodes[0])

    #               5
    #              /
    #             4
    #            /
    #           3
    #          /
    #         2
    #        /
    #       1
    def Test2(self):
        vals = [5, 4, 3, 2, 1]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, 1, -1),
            (1, 2, -1),
            (2, 3, -1),
            (3, 4, -1),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test2", nodes[0])

    # 1
    #  \
    #   2
    #    \
    #     3
    #      \
    #       4
    #        \
    #         5
    def Test3(self):
        vals = [1, 2, 3, 4, 5]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, -1, 1),
            (1, -1, 2),
            (2, -1, 3),
            (3, -1, 4),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test3", nodes[0])

    # 树中只有1个结点
    def Test4(self):
        vals = [1]
        nodes = binary_tree.createTreeNodesVec(vals)

        self.checkResult("Test4", nodes[0])

    # 树中没有结点
    def Test5(self):
        self.checkResult("Test5", None)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()
        self.Test4()
        self.Test5()


if __name__ == "__main__":
    test = Test()
    test.runTest()
